File: xls/dslx/token_parser.py
# Lint as: python3
#
# Copyright 2020 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

"""Base class for the parser that implements token peeking/popping."""

from typing import Text, Sequence, Union, Optional

from xls.dslx.parse_error import ParseError
from xls.dslx.python.cpp_pos import Pos
from xls.dslx.python.cpp_scanner import Keyword
from xls.dslx.python.cpp_scanner import Scanner
from xls.dslx.python.cpp_scanner import Token
from xls.dslx.python.cpp_scanner import TokenKind


class TokenParser(object):
  """Single-token lookahead parser (base class) that wraps a scanner."""

  def __init__(self, scanner: Scanner):
    self._scanner = scanner
    self._lookahead = None

  def _get_pos(self) -> Pos:
    if self._lookahead:
      return self._lookahead.span.start
    return self._scanner.pos

  def _at_eof(self) -> bool:
    return self._scanner.at_eof()

  def _peekt(self) -> Token:
    """Returns token that has been peeked at (non-destructively) from stream."""
    if self._lookahead is None:
      self._lookahead = self._scanner.pop()
      assert self._lookahead is not None

    return self._lookahead

  def _popt(self) -> Token:
    """Returns a token that has been popped destructively from token stream."""
    tok = self._peekt()
    self._lookahead = None
    return tok

  def _dropt(self) -> None:
    """Wraps _popt() to signify popping a token without needing the value."""
    self._popt()

  def _peekt_is(self, target: TokenKind) -> bool:
    assert not isinstance(target, Keyword), \
        'Not a token kind: {!r}'.format(target)
    return self._peekt().kind == target

  def _peekt_is_keyword(self, target: Keyword) -> bool:
    return self._peekt().is_keyword(target)

  def _peekt_is_identifier(self, target: Text) -> bool:
    return self._peekt().is_identifier(target)

  def _peekt_in(self, targets: Sequence[Union[TokenKind, Keyword]]) -> bool:
    tok = self._peekt()
    for target in targets:
      if isinstance(target, TokenKind) and tok.kind == target:
        return True
      if isinstance(target, Keyword) and tok.is_keyword(target):
        return True
    return False

  def _try_popt(self, target: TokenKind) -> bool:
    tok = self._peekt()
    if tok.kind == target:
      self._dropt()
      return True
    return False

  def _try_pop_keyword(self, target: Keyword) -> bool:
    tok = self._peekt()
    if tok.is_keyword(target):
      self._dropt()
      return True
    return False

  def _try_pop_identifier_token(self, target: Text) -> Optional[Token]:
    tok = self._peekt()
    if tok.is_identifier(target):
      return self._popt()
    return None

  def _try_pop_identifier(self, target: Text) -> bool:
    return bool(self._try_pop_identifier_token(target))

  def _popt_or_error(self,
                     target: TokenKind,
                     start: Optional[Token] = None,
                     context: Optional[Text] = None) -> Token:
    """Pops a token of the target kind 'target' or raises a ParseError."""
    tok = self._peekt()
    if tok.kind == target:
      return self._popt()
    if start is None:
      msg = "Expected '{}', got '{}'".format(target.value, tok.to_error_str())
    else:
      msg = ("Expected '{}' for construct starting with '{}' @ {}, "
             "got '{}'").format(target.value, start.to_error_str(), start.span,
                                tok.to_error_str())
    if context:
      msg += ': ' + context
    raise ParseError(tok.span, msg)

  def _dropt_or_error(self,
                      target: TokenKind,
                      start: Optional[Token] = None,
                      context: Optional[Text] = None) -> None:
    """Just a wrapper around _popt_or_error that doesn't return the token.

    This helps to signify that the intent was to drop the token in caller code
    vs 'forgetting' to do something with a popped token.

    Args:
      target: The token kind that we want to pop from the token stream; if this
        is not the kind we observe from the pop, we raise an error.
      start: An optional 'start of the parsing construct' token for use in error
        messages; e.g. pointing out the start of an open paren when we're trying
        to pop a corresponding closing paren.
      context: Context string to be used in reporting an error.

    Raises:
      ParseError: When the popped token is not of kind 'target'.
    """
    self._popt_or_error(target, start=start, context=context)

  def _pop_keyword_or_error(self,
                            keyword: Keyword,
                            context: Optional[Text] = None) -> Token:
    """Pops a token of target keyword and returns it or raises a ParseError."""
    tok = self._popt()
    if tok.is_keyword(keyword):
      return tok
    msg = 'Expected keyword \'{}\', got {}'.format(keyword.value.lower(),
                                                   tok.to_error_str())
    if context:
      msg += ': ' + context
    raise ParseError(tok.span, msg)

  def _drop_keyword_or_error(self, keyword: Keyword) -> None:
    self._pop_keyword_or_error(keyword)

File: xls/dslx/frontend/token_parser.cc
// Copyright 2021 The XLS Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "xls/dslx/frontend/token_parser.h"

#include <string>
#include <string_view>
#include <utility>

#include "absl/status/status.h"
#include "absl/status/statusor.h"
#include "absl/strings/str_cat.h"
#include "absl/strings/str_format.h"
#include "absl/types/span.h"
#include "xls/common/status/status_macros.h"
#include "xls/dslx/frontend/bindings.h"
#include "xls/dslx/frontend/pos.h"
#include "xls/dslx/frontend/token.h"

namespace xls::dslx {

absl::StatusOr<bool> TokenParser::PeekTokenIs(TokenKind target) {
  XLS_ASSIGN_OR_RETURN(const Token* tok, PeekToken());
  return tok->kind() == target;
}
absl::StatusOr<bool> TokenParser::PeekTokenIs(Keyword target) {
  XLS_ASSIGN_OR_RETURN(const Token* tok, PeekToken());
  return tok->IsKeyword(target);
}
absl::StatusOr<bool> TokenParser::PeekTokenIn(
    absl::Span<TokenKind const> targets) {
  XLS_ASSIGN_OR_RETURN(const Token* tok, PeekToken());
  for (TokenKind target : targets) {
    if (target == tok->kind()) {
      return true;
    }
  }
  return false;
}
absl::StatusOr<bool> TokenParser::PeekKeywordIn(
    absl::Span<Keyword const> targets) {
  XLS_ASSIGN_OR_RETURN(const Token* tok, PeekToken());
  if (tok->kind() != TokenKind::kKeyword) {
    return false;
  }
  for (Keyword target : targets) {
    if (target == tok->GetKeyword()) {
      return true;
    }
  }
  return false;
}

absl::StatusOr<Token> TokenParser::PopTokenOrError(TokenKind target,
                                                   const Token* start,
                                                   std::string_view context,
                                                   Pos* limit_pos) {
  XLS_ASSIGN_OR_RETURN(const Token* tok, PeekToken());
  if (limit_pos != nullptr) {
    *limit_pos = tok->span().limit();
  }
  if (tok->kind() == target) {
    return PopToken();
  }

  std::string msg;
  if (start == nullptr) {
    msg = absl::StrFormat("Expected '%s', got '%s'", TokenKindToString(target),
                          tok->ToErrorString());
    if (tok->IsKeyword(Keyword::kIf)) {
      msg +=
          " :: note that conditional syntax is `if test_expr { then_expr } "
          "else { else_expr }`";
    }
  } else {
    msg = absl::StrFormat(
        "Expected '%s' for construct starting with '%s' @ %s, got '%s'",
        TokenKindToString(target), start->ToErrorString(),
        start->span().ToString(), tok->ToErrorString());
  }
  if (!context.empty()) {
    absl::StrAppend(&msg, ": ", context);
  }
  return ParseErrorStatus(tok->span(), msg);
}

absl::Status TokenParser::DropTokenOrError(TokenKind target, const Token* start,
                                           std::string_view context,
                                           Pos* limit_pos) {
  XLS_ASSIGN_OR_RETURN(Token token,
                       PopTokenOrError(target, start, context, limit_pos));
  return absl::OkStatus();
}

absl::StatusOr<Token> TokenParser::PopKeywordOrError(Keyword keyword,
                                                     std::string_view context,
                                                     Pos* limit_pos) {
  XLS_ASSIGN_OR_RETURN(Token tok, PopToken());
  if (tok.IsKeyword(keyword)) {
    return std::move(tok);
  }
  std::string msg =
      absl::StrFormat("Expected keyword '%s', got %s'",
                      KeywordToString(keyword), tok.ToErrorString());
  if (!context.empty()) {
    absl::StrAppend(&msg, ": ", context);
  }
  return ParseErrorStatus(tok.span(), msg);
}

absl::Status TokenParser::DropKeywordOrError(Keyword target, Pos* limit_pos) {
  XLS_ASSIGN_OR_RETURN(Token token, PopKeywordOrError(target, /*context=*/"",
                                                      /*limit_pos=*/limit_pos));
  return absl::OkStatus();
}

}  // namespace xls::dslx
